pkg/tsdb/stackdriver/stackdriver.go

```diff
@@ -351,40 +351,64 @@ func (e *StackdriverExecutor) unmarshalResponse(res *http.Response) (Stackdriver
 }
 
 func (e *StackdriverExecutor) parseResponse(queryRes *tsdb.QueryResult, data StackdriverResponse, query *StackdriverQuery) error {
-	metricLabels := make(map[string][]string)
-	resourceLabels := make(map[string][]string)
-	var resourceTypes []string
-
-	for _, series := range data.TimeSeries {
-		if !containsLabel(resourceTypes, series.Resource.Type) {
-			resourceTypes = append(resourceTypes, series.Resource.Type)
-		}
-	}
+	labels := make(map[string][]string)
+	labels["resource.type"] = make([]string, 0)
 
 	for _, series := range data.TimeSeries {
 		points := make([]tsdb.TimePoint, 0)
+		seriesLabels := make(map[string]string)
 
 		defaultMetricName := series.Metric.Type
-		if len(resourceTypes) > 1 {
-			defaultMetricName += " " + series.Resource.Type
+
+		seriesLabels["resource.type"] = series.Resource.Type
+		if !containsLabel(labels["resource.type"], series.Resource.Type) {
+			labels["resource.type"] = append(labels["resource.type"], series.Resource.Type)
 		}
 
 		for key, value := range series.Metric.Labels {
-			if !containsLabel(metricLabels[key], value) {
-				metricLabels[key] = append(metricLabels[key], value)
+			if !containsLabel(labels["metric.label."+key], value) {
+				labels["metric.label."+key] = append(labels["metric.label."+key], value)
 			}
 			if len(query.GroupBys) == 0 || containsLabel(query.GroupBys, "metric.label."+key) {
 				defaultMetricName += " " + value
 			}
+			seriesLabels["metric.label."+key] = value
 		}
 
 		for key, value := range series.Resource.Labels {
-			if !containsLabel(resourceLabels[key], value) {
-				resourceLabels[key] = append(resourceLabels[key], value)
-			}
 			if containsLabel(query.GroupBys, "resource.label."+key) {
 				defaultMetricName += " " + value
 			}
+
+			if !containsLabel(labels["resource.label."+key], value) {
+				labels["resource.label."+key] = append(labels["resource.label."+key], value)
+			}
+			seriesLabels["resource.label."+key] = value
+		}
+
+		for labelType, labelTypeValues := range series.MetaData {
+			for labelKey, labelValue := range labelTypeValues {
+				output := []string{}
+				if values, ok := labelValue.([]interface{}); ok {
```

Comment:
Tried that intially, but it doesn't work. I guess go does only recognize at a slice are not aware of element types when casting. Anyway, refactored the code to use type switching instead of if statements. That should make it a bit more readable.

Reply:
LGTM